### PR TITLE
Port optimization pipeline to new pass manager

### LINKED ISSUE #2192: Port CodegenLLVM::optimize to new pass manager
Using LLVM's legacy PM for optimization pipeline was [deprecated in 13.0.0](https://releases.llvm.org/13.0.0/docs/ReleaseNotes.html#changes-to-the-llvm-ir) and the relevant functionality was scheduled to be removed after 14.0.0 (got delayed). The legacy PM for code generation (assembly, object file generation; mainly addPassesToEmitFile) is still supported.

Clang dropped legacy PM support in [D123609](https://reviews.llvm.org/D123609) (milestone: 15.0.0). I plan to remove instrumentation passes like gcov, PGO, and sanitizers (milestone: 15.0.0).

LLVM's legacy PM for optimization pipeline may be completely removed in 16.0.0 (`llvm::create*Pass`, `populateModulePassManager`, etc). llvm-project `clang/lib/CodeGen/BackendUtil.cpp` can serve as an example using the new pass manager.

In `src/ast/passes/codegen_llvm.cpp`, `PMB.populateModulePassManager(PM);` needs to be migrated.

### PR DESCRIPTION
**Pull request by MaskRay**
and use it for LLVM_VERSION_MAJOR >= 14.

Fixes #2192

---

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by MaskRay** [comment]
By adding a few lines this can be used with `LLVM_VERSION_MAJOR >= 12`, but it requires few more dispatches.
With 12 (tests require 12 at the moment,) `codegen-validator` in `make test` passes.

```patch
--- a/src/ast/passes/codegen_llvm.cpp
+++ b/src/ast/passes/codegen_llvm.cpp
@@ -9,4 +9,5 @@
 
 #include <llvm-c/Transforms/IPO.h>
+#include <llvm/Analysis/AliasAnalysis.h>
 #include <llvm/IR/Constants.h>
 #include <llvm/IR/LLVMContext.h>
@@ -3110,5 +3111,5 @@ void CodegenLLVM::optimize()
   assert(state_ == State::IR);
 
-#if LLVM_VERSION_MAJOR >= 14
+#if LLVM_VERSION_MAJOR >= 12
   PipelineTuningOptions pto;
   pto.LoopUnrolling = false;
@@ -3117,5 +3118,10 @@ void CodegenLLVM::optimize()
   pto.SLPVectorization = false;
 
-  llvm::PassBuilder pb(&orc_->getTargetMachine(), pto);
+  auto &tm = orc_->getTargetMachine();
+#if LLVM_VERSION_MAJOR >= 14
+  llvm::PassBuilder pb(&tm, pto);
+#else
+  llvm::PassBuilder pb(/*DebugLogging=*/false, &tm, pto);
+#endif
 
   // ModuleAnalysisManager must be destroyed first.
@@ -3125,4 +3131,10 @@ void CodegenLLVM::optimize()
   llvm::ModuleAnalysisManager mam;
 
+#if LLVM_VERSION_MAJOR < 14
+  // This is the default since LLVM 14.
+  llvm::AAManager aa = pb.buildDefaultAAPipeline();
+  fam.registerPass([&] { return std::move(aa); });
+#endif
+
   // Register all the basic analyses with the managers.
   pb.registerModuleAnalyses(mam);
@@ -3132,7 +3144,12 @@ void CodegenLLVM::optimize()
   pb.crossRegisterProxies(lam, fam, cgam, mam);
 
-  ModulePassManager mpm = pb.buildPerModuleDefaultPipeline(
-      llvm::OptimizationLevel::O3,
-      /*LTOPreLink=*/false);
+  ModulePassManager mpm;
+#if LLVM_VERSION_MAJOR >= 14
+  using OptimizationLevel = llvm::OptimizationLevel;
+#else
+  using OptimizationLevel = PassBuilder::OptimizationLevel;
+#endif
+  mpm = pb.buildPerModuleDefaultPipeline(OptimizationLevel::O3,
+                                         /*LTOPreLink=*/false);
   mpm.run(*module_, mam);
 
```

**Comment by viktormalik** [comment]
An embedded build is failing due to an LLVM error during parsing of one of the tools. Could you check if it is caused by this change?

**Comment by viktormalik** [comment]
> I think it is unrelated (this patch only affects llvm>=14), but I'll just rebase.

Hm, same error again. Master is passing, though, so this is very strange.

**Comment by MaskRay** [comment]
This is clearly a flaky bot. The patch does not affect LLVM 9.0.0: `-- Found LLVM 9.0.0: /usr/lib/llvm9/lib/cmake/llvm`

**Comment by viktormalik** [comment]
> This is clearly a flaky bot. The patch does not affect LLVM 9.0.0: `-- Found LLVM 9.0.0: /usr/lib/llvm9/lib/cmake/llvm`

I really seems so, but what is strange is that it is consistently failing for this PR only. I'll try to investigate it a bit before merging this.

File: src/ast/passes/codegen_llvm.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include <llvm/IR/LLVMContext.h>
 #include <llvm/IR/LegacyPassManager.h>
 #include <llvm/IR/Metadata.h>
+#include <llvm/Passes/PassBuilder.h>
```
Comment (viktormalik) [inline src/ast/passes/codegen_llvm.cpp:15]: The CI problem seems to stem from this include. Changing it to:
```suggestion
#if LLVM_VERSION_MAJOR >= 14
#include <llvm/Passes/PassBuilder.h>
#endif
```
makes the CI pass.

Comment (MaskRay) [reply src/ast/passes/codegen_llvm.cpp:16]: It's surprising that `llvm/Passes/PassBuilder.h` may cause differences. I have checked llvm 9 and do not find anything weird. Anyway, Added #if.

**Review by viktormalik — APPROVED** [review]

**Pull request merged**